File: JavaClient/src/main/java/com/security/crypto/Handshake/SessionHandler.java
package com.security.crypto.Handshake;

import com.security.crypto.Ciphers.AES.AesCBC;
import com.security.crypto.Ciphers.AES.AesECB;
import com.security.crypto.Configuration.CiphersForUse;
import com.security.crypto.Configuration.SupportedChipher;
import com.security.crypto.IOSocket.EstablishConnection;
import com.security.crypto.IOSocket.IOCallback;
import com.security.crypto.IOSocket.IOSynAck;
import com.security.crypto.KeyManager.KeyHandler;
import org.apache.commons.lang3.StringUtils;

import java.io.IOException;

public class SessionHandler {

    private EstablishConnection Session = null;
    private KeyHandler keystore = null;
    private AesECB Ecb = null;
    private AesCBC Cbc = null;

    private IOSynAck keyExchange = null;
    private IOCallback MessageExhange = null;
    private CiphersForUse ciphersforUse = null;
    private SupportedChipher cName;

    public SessionHandler(String connection) {
        this.Session = new EstablishConnection(connection);
        this.keystore = new KeyHandler();
        this.keyExchange = new DHkeyExchange(this.Session.getTransport(), this.keystore);
    }


    public void StartDHKeyExchange() {
        try {
            long elapsetime = System.currentTimeMillis();
            this.keyExchange.SendPlainMessage();
            long Execution_Time1 = (System.currentTimeMillis());
            System.out.println("---------------Execution Time1--------------------" + (Execution_Time1 - elapsetime));//60

            long Execution_Time2 = (System.currentTimeMillis());
            this.keyExchange.ReceiveServerCertificate();
            System.out.println("---------------Execution Time2--------------------" + (System.currentTimeMillis() - Execution_Time2));//1474378962584

            long Execution_Time3 = (System.currentTimeMillis());
            this.keyExchange.ResendCookieServer();
            System.out.println("---------------Execution Time3--------------------" + (System.currentTimeMillis() - Execution_Time3));//61

            long Execution_Time4 = (System.currentTimeMillis());
            this.keyExchange.SendPublicValue();
            System.out.println("---------------Execution Time4--------------------" + (System.currentTimeMillis() - Execution_Time4));//1474378962729

            long Execution_Time5 = (System.currentTimeMillis());
            this.keyExchange.ReceivePublicValue();
            System.out.println("---------------Execution Time5--------------------" + (System.currentTimeMillis() - Execution_Time5));//77

            long Execution_Time6 = (System.currentTimeMillis());
            this.keyExchange.SendCipherSuites();
            System.out.println("---------------Execution Time6--------------------" + (System.currentTimeMillis() - Execution_Time6));//778

            long Execution_Time7 = (System.currentTimeMillis());
            ciphersforUse = this.keyExchange.ReceiveCipherSuites();
            System.out.println("---------------Execution Time7--------------------" + (Execution_Time7 - System.currentTimeMillis()));
            System.out.println("---------------Sum up Time------------------------ " + (System.currentTimeMillis() - elapsetime));//900 ms total

            this.MessageExhange = new IOMessageExhange(this.Session.getTransport(), this.keystore, ciphersforUse);
            StoreCipher();
        } catch (IOException e) {
            e.printStackTrace();
            this.ConnectionClose();
            return;
        } catch (Exception e) {
            e.printStackTrace();
            this.ConnectionClose();
            return;
        }
    }

    private void StoreCipher() {
        if (StringUtils.containsIgnoreCase(this.ciphersforUse.getCipherAlgorithm(), "ECB")) {
            this.cName = SupportedChipher.ECB;
            this.Ecb = new AesECB();
        } else {
            this.cName = SupportedChipher.CBC;
            this.Cbc = new AesCBC();
        }
    }

    public void SendSecureMessage(String Message) {
        try {
            switch (cName) {
                case ECB:
                    this.MessageExhange.SendDHEncryptedMessage(Message, this.Ecb);
                    break;
                case CBC:
                    this.MessageExhange.SendDHEncryptedMessage(Message, this.Cbc);
                    break;
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public String ReceiveSecureMessage() {
        try {
            switch (cName) {
                case ECB:
                    return this.MessageExhange.ReceiveDHEncryptedMessage(this.Ecb);

                case CBC:
                    return this.MessageExhange.ReceiveDHEncryptedMessage(this.Cbc);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    public void ConnectionClose() {
        try {
            Session.getTransport().close();
        } catch (IOException ex) {
            System.exit(1);
        }
    }

}

File: PythonClient/Handshake/SessionHandler.py
import time

from Crypto.Cipher import AES

from Ciphers.Symmetric.AesCBC import CBC
from Ciphers.Symmetric.AesECB import ECB
from Configuration import Properties
from Handshake.DHkeyExchange import DiffeHelmanExhange
from Handshake.IOMessageExchange import MessageExchange
from IOSocket.SSLSocket import SSLSocket
from IOSocket.Socket import PlainSocket
from KeyManager.KeyHandler import KeyHandle

class HandleSession:
    def __init__(self, ConnType):
        self._ConnType = ConnType
        self._Keystore = KeyHandle()
        self.__ecb = None
        self.__cbc = None
        self.__EstablishConn()

    def __StartExhangeKey__(self):
        elapsetime = int(round(time.time() * 1000))

        self._DHExchange._SynAck__SendPlainMessage()
        Execution_Time1 = int(round(time.time() * 1000))
        print("---------------Execution Time1--------------------" + str(Execution_Time1 - elapsetime))

        self._DHExchange._SynAck__ReceiveServerCertificate()
        Execution_Time2 = int(round(time.time() * 1000))
        print("---------------Execution Time2--------------------" + str(Execution_Time2 - Execution_Time1))

        self._DHExchange._SynAck__ResendCookieServer()
        Execution_Time3 = int(round(time.time() * 1000))
        print("---------------Execution Time3--------------------" + str(Execution_Time3 - Execution_Time2))

        self._DHExchange._SynAck__SendPublicValue()
        Execution_Time4 = int(round(time.time() * 1000))
        print("---------------Execution Time4--------------------" + str(Execution_Time4 - Execution_Time3))

        self._DHExchange._SynAck__ReceivePublicValue()
        Execution_Time5 = int(round(time.time() * 1000))
        print("---------------Execution Time5--------------------" + str(Execution_Time5 - Execution_Time4))

        self._DHExchange._SynAck__SendCipherSuites()
        Execution_Time6 = int(round(time.time() * 1000))
        print("---------------Execution Time6--------------------" + str(Execution_Time6 - Execution_Time5))

        SumUpTime = int(round(time.time() * 1000))
        self._ciphersforUse = self._DHExchange._SynAck__ReceiveCipherSuites()
        print("---------------Execution Time6--------------------" + str(SumUpTime - Execution_Time6))
        print("---------------DHkeys Sucessfuly Changed--------------------" + str(SumUpTime - elapsetime))
        self._MessageExhange = MessageExchange(self._Plain, self._Keystore, self._ciphersforUse)
        self.StoreCipher()

    def StoreCipher(self):
        if self._ciphersforUse.CipherAlgorithm == AES.MODE_ECB:
            self.__ecb = ECB()
        elif self._ciphersforUse.CipherAlgorithm == AES.MODE_CBC:
            self.__cbc = CBC()

    def __SendSecurMessage__(self, Message):
        if self.__ecb is not None:
            self._MessageExhange._Callback__SendDHEncryptedMessage(Message, self.__ecb)
        elif self.__cbc is not None:
            self._MessageExhange._Callback__SendDHEncryptedMessage(Message, self.__cbc)

    def __ReceiveSecurMessage__(self):
        if self.__ecb is not None:
            return self._MessageExhange._Callback__ReceiveDHEncryptedMessage(self.__ecb)
        elif self.__cbc is not None:
            return self._MessageExhange._Callback__ReceiveDHEncryptedMessage(self.__cbc)

    def __EstablishConn(self):
        if str(self._ConnType).__eq__(str(Properties.Plain)):
            self._Plain = PlainSocket()
            self._DHExchange = DiffeHelmanExhange(self._Plain, self._Keystore)
        elif str(self._ConnType).__eq__(str(Properties.Secure)):
            self._SSLSock = SSLSocket()
        else:
            raise ValueError("Not Compatible !!")

    def __Close__(self):
        try:
            self._Plain.__Close__()
        except:
            self._SSLSock.__Close__()
